### PR TITLE
Rule-Only Drained and Bound AP

### PR DESCRIPTION
**Pull request by Tinkastrong**
~~## There is also a rule change in here~~
~~To emphasize: i _**personally**_ think it is a bug of the stat squish PDF to not make sure you have an increased AP growth. It seems more in spirit with the Variant Rule that grants everything fast to acutally grant EVERYTHING faster.~~

---

[Screencast from 2023-10-29 21-46-32.webm](https://github.com/dylanpiera/Foundry-Pokemon-Tabletop-United-System/assets/149251521/bad1ff38-35c2-4cf4-9857-c5e35317954c)

---
- Feature
  - Max AP Mod - Extras Tab of Trainers have new Max AP Modifier - Max AP can be changed through rules at path `system.modifiers.maxAp.mod`
  - Level Squish
    - When using the Trainer Revamp (Level 25) Variant Rule, Max AP increase every 2.5 level instead of every 5 
      - AP increase before: 5, 10, 15, ... 
      - AP increase  after: 3,  5,  8, 10, ...

### CODE REVIEW

**Comment by Tinkastrong** [comment]
Got some support for the rule change part (variant role squish given AP every 2.5 level)
https://discord.com/channels/748601513835888682/1168285583173488690

File: src/module/actor/character/document.js [hunk 1]
```diff
@@ -148,7 +148,7 @@ class PTUTrainerActor extends PTUActor {
                 + (game.settings.get("ptu", "variant.trainerRevamp") ? (system.level.current >= 25 ? 1 : 0) : 0)
         }
 
-        system.ap.max = 5 + Math.floor(system.level.current / 5);
+        system.ap.max = 5 + Math.floor(system.level.current / (game.settings.get("ptu", "variant.trainerRevamp") ? 2.5 : 5)) + (system.modifiers?.maxAp?.total ? system.modifiers?.maxAp?.total  : 0);
```
Comment (dylanpiera) [inline src/module/actor/character/document.js:151]:
```suggestion
        system.ap.max = 5 + Math.floor(system.level.current / (game.settings.get("ptu", "variant.trainerRevamp") ? 2.5 : 5)) + (system.modifiers?.maxAp?.total ?? 0);
```
Use [Nullish Coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) instead for conciser code

File: static/templates/actor/trainer-sheet-compact.hbs
Comment (dylanpiera) [inline static/templates/actor/trainer-sheet-compact.hbs:1]: You ran a formatter on the HBS file that is mucking up the syntax. Please only commit actual changes as I can't compare what you have changed in this file.

**Review by dylanpiera — CHANGES_REQUESTED** [review]
Will have to make a decision on the condensed ruling for max AP (get it more frequently vs get double the amount on standard levels).

Besides that you said you added a Max AP field in the extra tab, what is the use of this? Is this purely for homebrew or are there feats/edges/etc. in PTR Core that modify Max AP and are currently un-automatable?

The extra tab is being phased out so I'd prefer not adding new values here.

**Comment by Tinkastrong** [comment]
>Will have to make a decision on the condensed ruling for max AP (get it more frequently vs get double the amount on standard levels).

:+1: 

> Besides that you said you added a Max AP field in the extra tab, what is the use of this? Is this purely for homebrew or are there feats/edges/etc. in PTR Core that modify Max AP and are currently un-automatable?

Bound/Drained AP can not be reflected on a character sheet. While not too bad for bound AP within a scene, it is a hindrance wrt Drained AP on long adventuring days as well as Bound AP people might wish to keep bound for a longer time - Fortress >  Stalwart Bastion.

> Use [Nullish Coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) instead for conciser code

Thanks for advice, greatly appreciated.

> You ran a formatter on the HBS file that is mucking up the syntax. Please only commit actual changes as I can't compare what you have changed in this file.

Sorry, that is a total f*up on my end. Should not have happened.

> The extra tab is being phased out so I'd prefer not adding new values here.

Do you already have a plan worth sharing on how to re-distribute the Fields from that tab elsewhere?

File: static/templates/actor/trainer-sheet-compact.hbs
Comment (Tinkastrong) [reply static/templates/actor/trainer-sheet-compact.hbs:1]: sry, undone

**Comment by Tinkastrong** [comment]
Should there still be a way to manually edit Drained and Bound AP? I assume yes, but would like confirmation.

**Comment by Tinkastrong** [comment]
Removed change to AP scaling when variantRule for level squish is enabled.

**Comment by Tinkastrong** [comment]
There is now an ActionPointRuleElement. It handles reducing and returning current AP based on bound, drained and unbound AP. Works fine for all changes made through Effects and such.

There are some TODOs that need guidance:
- There is currently no Info-on-Hover like for mod fields in the extra Tab. I do not know how easy or hard it is to use or adapt the `static/templates/actor/trainer-sheet-compact.hbs` to make it show all rules applying to the newly introduced synthetic.
- I do not know how this can be properly adopted to allow a manually entered value from somewhere. The `actor.system.modifiers` explicitly divide `value` and `mod`, the manually entered value and the sum of all automatic mods. If a manual value for drained/bound AP is to exist somewhere in the sheet, it bypasses the code withtin ActionPointRuleElement, and thus needs it's own logic for reducing and returning AP when changed. The `PTUActor`, where the `actor.system.modifiers` are handled, does afaik not have a reference to a previous state of itself, thus not allowing the see changes in the manual value. One **idea** would be for the manual drained/bound fields to go into **a ActionPointRuleElement attached to the Actor itself**. While this solves all previously mentioned issues, I do not know how good how a solution this would be.

**Comment by dylanpiera** [comment]
imo they don't need fields if we have automation for them; we can have generic "drain 1/2 and bind 1/2" AP effects in the compendium that people can drag in for items that don't have them innately added.

**Comment by dylanpiera** [comment]
Also you can look at the level up points hover field 😉

**Comment by Tinkastrong** [comment]
[Screencast from 2023-11-02 12-15-09.webm](https://github.com/dylanpiera/Foundry-Pokemon-Tabletop-United-System/assets/149251521/8cada938-abf5-4029-b822-40a52b9bcdeb)

**Comment by Tinkastrong** [comment]
Study Manual Tier 2 Effect with Choice Set for Skill.
[fvtt-Item-study-manual-tier-2-FgAGsZsHpNUNJHrY.json](https://github.com/dylanpiera/Foundry-Pokemon-Tabletop-United-System/files/13238302/fvtt-Item-study-manual-tier-2-FgAGsZsHpNUNJHrY.json)

File: src/module/rules/rule-element/ap.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        const maxApBeforeRule = (Number(this.actor.system.ap.max) || 0) + this.drainedValue + this.boundValue;
+        const currentApAfterRule = Math.clamped(currentApBeforeRule - this.drainedValue - this.boundValue, 0, maxApBeforeRule - this.drainedValue - this.boundValue);
+        mergeObject(actorUpdates, {
+            "system.ap.current": currentApAfterRule
+        });
+    }
+
+    afterPrepareData() {
+        if(this.ignored) return;
+        super.afterPrepareData()
```
Comment (dylanpiera) [inline src/module/rules/rule-element/ap.js:35]:
```suggestion
```
Doesn't do anything; don't add code that in theory later down the road could cause unforseen consequences.

File: template.json [hunk 3]
```diff
@@ -468,8 +468,9 @@
         "injuries": 0,
         "max": 0
       },
+
       "ap": {
-        "value": 5
+        "current": 5
       },
```
Comment (dylanpiera) [inline template.json:474]: You cannot just make a change to template.json and expect it to work. 

Everyone that updates their foundry to this version all their trainer actors would now have an `ap.value` field set to whatever it was set and a new field created `ap.current` with value 5. A table that left in the middle of a session would suddenly have all their AP reset to 5.

Renaming a value like this should only be done if it's actually necessary, and then a migration script needs to be written as well.

Please change it back to `value` as well as every reference from `current` back to `value`, or ping me on discord if you have a good reason for changing it 😉

**Review by dylanpiera — CHANGES_REQUESTED** [review]

**Comment by Tinkastrong** [comment]
TODO test after reverting value - current - naming, gimme 10 minutes

**Comment by Tinkastrong** [comment]
@dylanpiera Do you have any fix how to make this less... *utilitarian*
image

**Comment by dylanpiera** [comment]
Merging it into the Beta 16 branch and will fix the styling issue there (and any bug I may encounter)

**Pull request merged**